jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java

```diff
@@ -5599,6 +5600,66 @@ static MethodHandle collectReturnValue(MethodHandle target, MethodHandle filter)
 		}
 		return buildTransformHandle(new CollectReturnHelper(target, filter), resultType);
 	}
+	
+	/**
+	 * Return an array of checked exceptions for a MethodHandle.
+	 * 
+	 * @param mh A MethodHandle to scan for checked exceptions.
+	 * 
+	 * @return an array of checked exception classes if any or null.
+	 * 
+	 * @throws InternalError for an error.
+	 */
+	static Class<?>[] checkedExceptions(MethodHandle mh) {
```

Comment:
This is more complicated than it needs to be as it tries to lazy init the lists.  Something like the code below is more explicit that it's processing the MH chain as a worklist until it has found all the ExceptionTypes in the MH chain:
```java
	static Class<?>[] getExceptions(MethodHandle mh) {
		if (mh == null) {
			return null;
		}
		Deque<MethodHandle> workList = = new ArrayDeque<>();
		relatedMHsList.push(mh);
		List<Class<?>> exceptionTypesList = new ArrayList<>(4);

		while (!workList.isEmpty()) {
			mh = workList.pop();
			if (mh instanceof PrimitiveHandle) {
				Class<?>[] exceptionTypes = null;
				MethodHandleInfoImpl mhi = new MethodHandleInfoImpl((PrimitiveHandle)mh);
				if (mhi.isConstructor()) {
					exceptionTypes = mhi.reflectAs(Constructor.class, Lookup.IMPL_LOOKUP).getExceptionTypes();
				} else if (mhi.isMethod()) {
					exceptionTypes = mhi.reflectAs(Method.class, Lookup.IMPL_LOOKUP).getExceptionTypes();
				} else if (mhi.isField()) {
					/* Field reference kind has no checked exceptions. */
				} else {
					/*[MSG "K0686", "Unknown reference kind: '{0}'"]*/
					throw new InternalError(com.ibm.oti.util.Msg.getString("K0686", mhi.getReferenceKind())); //$NON-NLS-1$
				}
				Collections.addAll(exceptionTypesList, exceptionTypes);
			} else {
				MethodHandle[] relatedMHs = mh.getRelatedMHs();
				workList.addAll(relatedMHs);
			}
		}
		Class<?>[] exceptionTypesArray = null;
		if (!exceptionTypesList.isEmpty()) {
			exceptionTypesArray = exceptionTypesList.toArray(new Class<?>[0]);
		}
		return exceptionTypesArray;
	}
```

I left `exceptionTypesList` as a List though it may be better as a `HashSet` to avoid repeated processing of the same exception classes.  An even better approach may be to check the exception types as they are fetched to see if any are checked or not.  The caller only cares for a boolean response, correct?

Reply:
Correct, caller only cares about a boolean response. Updated to return a boolean type and process the exception types are they are fetched. I continue to use an `ArrayList` for supporting `addRelatedMHs`. Removing the last element from an `ArrayList` is inexpensive. Also, accounted for other explicit changes.